### PR TITLE
Tests with signatures, fixes #1074

### LINKED ISSUE #1074: BLS and testing
Decided I wanted to get this out to explain the current state of testing, and **collect feedback** (implementers please comment) on what you need from testing, and your feelings about BLS usage in tests.

# BLS and testing

The two pain-points to get a pretty (and large) set of test-vectors out for clients are:
- BLS Signature creation
- BLS Signature verification

And side-issue, but easily resolved:
*efficient creation of a genesis state*:
When BLS functionality is implemented in test-code (creation of signed deposits, and verification).
Solution would be to either cache it, or create it directly, without going through the spec functions (current temporary solution on experiment branch).

## Status

Talking about the status on  [`spectest-deco` PR 1052](https://github.com/ethereum/eth2.0-specs/pull/1052) here, based on the  `v06x` branch, where we are developing 0.6 improvements. (to be merged back into dev later)

### The testing pipeline currently looks like:

- py-spec, calls BLS stub
- test-helpers, don't create self-signed objects with valid signatures
- py-test code, unified with test-vector-creation (see [PR 1052](https://github.com/ethereum/eth2.0-specs/pull/1052))
- py-test runner to run spec-tests, purely for assertions
- test-generator running the spec-tests, passing `generator_mode=true` to each of them, making them output a test-vector.

### Pytests status:

- move from `tests/` to `eth2spec/test`, i.e. part of package
    - removed use of `pytest`
    - annotated with `@spec_test` or similar (see PR 1052)
- as part of test-generation effort, yay for shared effort:
    - expanded in block-operation testing: [coverage checklist here](https://github.com/ethereum/eth2.0-specs/issues/927)
    - slightly faster, less deep-copies
- stuck on BLS stub (no sig creation/verification)

### Test-generation status:

- BLS, SSZ-generic, SSZ-static, shuffling test generators still all in place and up to date (`v06x` branch)
- `operations` test-gen uses test-package ability to output test-vectors for each test-case
    - but no valid signatures
    - lack of a definition how to handle this signature problem as a test-consumer
        - there are no signature-related testcases
        - turning BLS off would effectively let you check conformance, but it's hacky, and not remotely a good practice to have even an option for...
    - it's approx. ~140MB worth (iirc) of yaml encoded state-transitions, covering many edge-cases. Worth to get in the hands of implementers quick.
- `sanity` tests updated and can be cleanly used for test-generation, but requires more work to define the format of the test-vectors, as they is more variety.
- `epoch` processing tests also updated, also can be used, not as complete as block-processing, lower priority.

## Possible ways forward:

- Simple but hacky: "turn BLS off for testing"
- No "BLS off", BLS ON on client side, but only partially on spec side. Rely on signature verification not being hit before anything else during testing
    - valid test cases generated with valid signatures
    - invalid test cases marked: does it error because of BLS? And runners should check the reason for aborting processing: if it doesn't match, the test should fail. Now these pytests don't need full BLS update work, and can be released somewhat quicker
- "BLS on", more work (~1 week)
    - slower on test-generation, but we get the best kind of test-vectors: correct, BLS verification ON.
    - blocker: what if a test case fails because of a signature error (test setup not creating the sig correctly), instead of a real assertion case. Spec will look correct, passes tests, but things are not right. We need to mark Sig-verification errors distinctly, so we can catch these problems when we turn BLS on in the pyspec. How: instead of `assert verify_...`, just `verify_...`, and make it raise a special `BLSVerificationError` (or something like that)
    - We likely still want to mark tests as "signature related" or not, so implementers can catch it easily if their code is not aborting properly before signature verification, to assure invalid inputs are not costly.

A work-in-progress introduction of actual full BLS usage in the pytests is started here: [`tests-with-sigs` branch](https://github.com/ethereum/eth2.0-specs/tree/tests-with-sigs)

Suggestions welcome.

### PR DESCRIPTION
**Pull request by protolambda**
Made the pytests respect BLS signatures. This enables us to use the tests to generate properly BLS-signed test-vectors with.
Note: there are a few exceptions:
- finality tests are too slow (many transitions), hence they run with disabled BLS. I'm not sure how/when we include these in the test-vectors for clients
- there are a few sanity tests that I just disabled, as they cover long-range (empty slots) block transitions,  costing lots of hashing.

The block-operations and epoch transitions are fully signed now :tada:

TODO:
- [x] Invalid-signature tests for block operations
- [ ] ~~Cover more of epoch transition with new tests~~ future PR
- [x] Call the tests in generator-mode to generate test-vectors. (epoch transition generator in future PR)
- [ ] ~~Implement some basic SSZ caching, to speed up tests (the hashes of fields of empty blocks being cached would help a lot! Tests typically change only a part of the block).~~ CI runs fast enough, SSZ caching to be implemented in separate PR

### CODE REVIEW

**Comment by djrtwo** [comment]
Ah, looks like this branch doesn't have the changes from #1052 integrated. Is the intention to modify `generate_from_tests` to override the default of BLS being off to on?

**Comment by hwwhww** [comment]
IMO it's safer to make it more explicit by removing the default value `False` from `signed` parameter.

File: test_libs/pyspec/eth2spec/test/block_processing/test_process_block_header.py [hunk 1]
```diff
@@ -42,23 +42,24 @@ def run_block_header_processing(state, block, valid=True):
...
+def test_success_block_header(state):
+    block = build_empty_block_for_next_slot(state, signed=True)
     yield from run_block_header_processing(state, block)
 
 
 @spec_state_test
-def test_invalid_slot(state):
-    block = build_empty_block_for_next_slot(state)
+def test_invalid_slot_block_header(state):
+    block = build_empty_block_for_next_slot(state, signed=True)
```
[the remark this answers is not in the document]
Comment (protolambda) [reply test_libs/pyspec/eth2spec/test/block_processing/test_process_block_header.py]: Technically, yes. The problem with this test is that signing future blocks is a poor experience: we have to calculate the proposer index, which is slow due to transitions being ran on a copied state for the calculation. In this case just signing may still be fine. I'll change it :+1:

**Comment by protolambda** [comment]
@djrtwo 

> Ah, looks like this branch doesn't have the changes from #1052 integrated.

It's branched off of that, so it should. You mean the proposal for BLS test-improvements? This PR is in draft mode mostly because:
- want to implement some invalid-signature tests
- need to go over the BLS things listed in this comment: https://github.com/ethereum/eth2.0-specs/issues/1074#issuecomment-491992791 to see what we really want, and what is better to adjust. (i.e. test-markers for BLS set to always ON)

> Is the intention to modify generate_from_tests to override the default of BLS being off to on?

Ideally I think BLS should be ON, if pytests were fast enough. But we already discussed this, and it will be off for CI where possible, but ON where needed + in generated vectors.

For now I just like the explicitness, as it helped me tracking what needs to be updated, and finding testing bugs (things not being signed). If things are stable we may go for implicit signed True/False, but I don't think it's a netto win now.

File: test_libs/pyspec/eth2spec/test/block_processing/test_process_deposit.py [hunk 2]
```diff
@@ -34,21 +34,27 @@ def run_deposit_processing(state, deposit, validator_index, valid=True):
 
     yield 'post', state
 
-    if validator_index < pre_validator_count:
-        # top-up
+    if non_effective:
```
[the remark this answers is not in the document]
Comment (protolambda) [reply test_libs/pyspec/eth2spec/test/block_processing/test_process_deposit.py]: I prefer defaults to be false / zero, it's great when you can guess behavior and it just works. Negative boolean is a small trade-off here. I'll rename it to ineffective.

[the remark this answers is not in the document]
Comment (protolambda) [reply test_libs/pyspec/eth2spec/test/block_processing/test_process_deposit.py]: Hmm, nvm, True may be more consistent still, since we already have `valid=True`. :thinking:

[the remark this answers is not in the document]
Comment (protolambda) [reply test_libs/pyspec/eth2spec/test/block_processing/test_process_deposit.py]: Fixed in 58fd712
